File: src/cpp_pkg/src/benchmark_node.cpp
#include <algorithm>
#include <atomic>
#include <chrono>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <mutex>
#include <numeric>
#include <vector>
#include <thread>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp/executors/multi_threaded_executor.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "std_msgs/msg/string.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"
#include "example_interfaces/action/fibonacci.hpp"

using namespace std::chrono;
using Fibonacci = example_interfaces::action::Fibonacci;
using GoalHandleFib = rclcpp_action::ServerGoalHandle<Fibonacci>;

struct Stats {
    double avg_ns;
    double min_ns;
    double max_ns;
    double p50_ns;
    double p95_ns;
    double p99_ns;
    double throughput;
};

Stats compute_stats(std::vector<int64_t>& latencies, double elapsed_s) {
    std::sort(latencies.begin(), latencies.end());
    size_t n = latencies.size();
    double sum = std::accumulate(latencies.begin(), latencies.end(), 0.0);
    return Stats{
        sum / n,
        static_cast<double>(latencies.front()),
        static_cast<double>(latencies.back()),
        static_cast<double>(latencies[n / 2]),
        static_cast<double>(latencies[static_cast<size_t>(n * 0.95)]),
        static_cast<double>(latencies[static_cast<size_t>(n * 0.99)]),
        n / elapsed_s
    };
}

void print_json_stats(const std::string& name, const Stats& s) {
    std::cout << "  \"" << name << "\": {"
              << "\"avg_us\": " << s.avg_ns / 1000.0
              << ", \"min_us\": " << s.min_ns / 1000.0
              << ", \"max_us\": " << s.max_ns / 1000.0
              << ", \"p50_us\": " << s.p50_ns / 1000.0
              << ", \"p95_us\": " << s.p95_ns / 1000.0
              << ", \"p99_us\": " << s.p99_ns / 1000.0
              << ", \"throughput\": " << s.throughput
              << "}";
}

void benchmark_topic(rclcpp::Node::SharedPtr node, int iterations) {
    auto pub = node->create_publisher<std_msgs::msg::String>("bench_topic_cpp", rclcpp::QoS(100));
    std::vector<int64_t> latencies;
    latencies.reserve(iterations);

    auto sub = node->create_subscription<std_msgs::msg::String>(
        "bench_topic_cpp", rclcpp::QoS(100),
        [&latencies](const std_msgs::msg::String::SharedPtr msg) {
            auto now = high_resolution_clock::now().time_since_epoch();
            auto now_ns = duration_cast<nanoseconds>(now).count();
            auto sent_ns = std::stoll(msg->data);
            latencies.push_back(now_ns - sent_ns);
        });

    auto start = high_resolution_clock::now();
    for (int i = 0; i < iterations; ++i) {
        auto msg = std_msgs::msg::String();
        auto now_ns = duration_cast<nanoseconds>(
            high_resolution_clock::now().time_since_epoch()).count();
        msg.data = std::to_string(now_ns);
        pub->publish(msg);
        rclcpp::spin_some(node);
        std::this_thread::sleep_for(microseconds(10));
    }
    for (int i = 0; i < 50; ++i) {
        rclcpp::spin_some(node);
        std::this_thread::sleep_for(milliseconds(1));
    }
    auto end = high_resolution_clock::now();
    double elapsed_s = duration<double>(end - start).count();

    if (!latencies.empty()) {
        auto stats = compute_stats(latencies, elapsed_s);
        print_json_stats("topic", stats);
    } else {
        std::cout << "  \"topic\": {\"error\": \"no messages received\"}";
    }
}

void benchmark_service(rclcpp::Node::SharedPtr node, int iterations) {
    auto srv = node->create_service<example_interfaces::srv::AddTwoInts>(
        "bench_srv_cpp",
        [](const example_interfaces::srv::AddTwoInts::Request::SharedPtr req,
           example_interfaces::srv::AddTwoInts::Response::SharedPtr resp) {
            resp->sum = req->a + req->b;
        });

    auto client = node->create_client<example_interfaces::srv::AddTwoInts>("bench_srv_cpp");
    while (!client->wait_for_service(seconds(1))) {
        rclcpp::spin_some(node);
    }

    std::vector<int64_t> latencies;
    latencies.reserve(iterations);

    auto start = high_resolution_clock::now();
    for (int i = 0; i < iterations; ++i) {
        auto req = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        req->a = i;
        req->b = 1;
        auto t0 = high_resolution_clock::now();
        auto future = client->async_send_request(req);
        rclcpp::spin_until_future_complete(node, future, seconds(5));
        auto t1 = high_resolution_clock::now();
        latencies.push_back(duration_cast<nanoseconds>(t1 - t0).count());
    }
    auto end = high_resolution_clock::now();
    double elapsed_s = duration<double>(end - start).count();

    auto stats = compute_stats(latencies, elapsed_s);
    print_json_stats("service", stats);
}

void benchmark_action(rclcpp::Node::SharedPtr node, int fib_order) {
    auto action_server = rclcpp_action::create_server<Fibonacci>(
        node, "bench_action_cpp",
        [](const rclcpp_action::GoalUUID&, std::shared_ptr<const Fibonacci::Goal>) {
            return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
        },
        [](const std::shared_ptr<GoalHandleFib>) {
            return rclcpp_action::CancelResponse::ACCEPT;
        },
        [](const std::shared_ptr<GoalHandleFib> gh) {
            std::thread([gh]() {
                auto goal = gh->get_goal();
                auto fb = std::make_shared<Fibonacci::Feedback>();
                auto result = std::make_shared<Fibonacci::Result>();
                int prev = 0, curr = 1;
                for (int i = 0; i < goal->order; ++i) {
                    fb->sequence.push_back(curr);
                    gh->publish_feedback(fb);
                    int next = prev + curr;
                    prev = curr;
                    curr = next;
                }
                result->sequence = fb->sequence;
                gh->succeed(result);
            }).detach();
        });

    auto action_client = rclcpp_action::create_client<Fibonacci>(node, "bench_action_cpp");
    action_client->wait_for_action_server(seconds(5));

    auto goal = Fibonacci::Goal();
    goal.order = fib_order;
    int feedback_count = 0;

    auto opts = rclcpp_action::Client<Fibonacci>::SendGoalOptions();
    opts.feedback_callback = [&feedback_count](auto, auto) { ++feedback_count; };

    auto t0 = high_resolution_clock::now();
    auto goal_future = action_client->async_send_goal(goal, opts);
    rclcpp::spin_until_future_complete(node, goal_future, seconds(10));
    auto goal_handle = goal_future.get();
    if (goal_handle) {
        auto result_future = action_client->async_get_result(goal_handle);
        rclcpp::spin_until_future_complete(node, result_future, seconds(10));
    }
    auto t1 = high_resolution_clock::now();
    double elapsed_us = duration_cast<microseconds>(t1 - t0).count();

    std::cout << "  \"action\": {"
              << "\"total_us\": " << elapsed_us
              << ", \"feedback_count\": " << feedback_count
              << ", \"fib_order\": " << fib_order
              << "}";
}

void benchmark_param(rclcpp::Node::SharedPtr node, int iterations) {
    node->declare_parameter("bench_int", 0);
    node->declare_parameter("bench_str", "init");

    auto t0 = high_resolution_clock::now();
    for (int i = 0; i < iterations; ++i) {
        node->set_parameter(rclcpp::Parameter("bench_int", i));
    }
    auto t1 = high_resolution_clock::now();
    double write_elapsed_s = duration<double>(t1 - t0).count();

    auto t2 = high_resolution_clock::now();
    volatile int64_t sink = 0;
    for (int i = 0; i < iterations; ++i) {
        sink = node->get_parameter("bench_int").as_int();
    }
    auto t3 = high_resolution_clock::now();
    double read_elapsed_s = duration<double>(t3 - t2).count();

    std::cout << "  \"param\": {"
              << "\"write_ops_per_sec\": " << iterations / write_elapsed_s
              << ", \"read_ops_per_sec\": " << iterations / read_elapsed_s
              << ", \"iterations\": " << iterations
              << "}";
}

void benchmark_concurrency(int num_threads, int msgs_per_thread,
                           int calls_per_thread, int param_ops_per_thread) {
    std::cout << "  \"concurrency\": {\"thread_count\": " << num_threads;

    // --- Multi-publisher topic ---
    {
        auto node = rclcpp::Node::make_shared("cpp_bench_conc_topic");
        std::mutex lat_mutex;
        std::vector<int64_t> latencies;
        latencies.reserve(num_threads * msgs_per_thread);

        auto sub = node->create_subscription<std_msgs::msg::String>(
            "bench_conc_topic_cpp", rclcpp::QoS(2000),
            [&](const std_msgs::msg::String::SharedPtr msg) {
                auto now_ns = duration_cast<nanoseconds>(
                    high_resolution_clock::now().time_since_epoch()).count();
                auto sent_ns = std::stoll(msg->data);
                std::lock_guard<std::mutex> lk(lat_mutex);
                latencies.push_back(now_ns - sent_ns);
            });

        auto executor = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
        executor->add_node(node);
        std::thread spin_thread([&executor]() { executor->spin(); });

        std::vector<std::thread> workers;
        auto t0 = high_resolution_clock::now();
        for (int t = 0; t < num_threads; ++t) {
            workers.emplace_back([&node, msgs_per_thread]() {
                auto pub = node->create_publisher<std_msgs::msg::String>(
                    "bench_conc_topic_cpp", rclcpp::QoS(2000));
                for (int i = 0; i < msgs_per_thread; ++i) {
                    auto msg = std_msgs::msg::String();
                    msg.data = std::to_string(duration_cast<nanoseconds>(
                        high_resolution_clock::now().time_since_epoch()).count());
                    pub->publish(msg);
                }
            });
        }
        for (auto& w : workers) w.join();
        std::this_thread::sleep_for(milliseconds(300));
        auto t1 = high_resolution_clock::now();
        double elapsed_s = duration<double>(t1 - t0).count();

        executor->cancel();
        spin_thread.join();

        size_t received = latencies.size();
        double avg_lat = 0;
        if (!latencies.empty()) {
            double sum = std::accumulate(latencies.begin(), latencies.end(), 0.0);
            avg_lat = sum / latencies.size() / 1000.0;
        }

        std::cout << ", \"topic_multi_pub\": {"
                  << "\"total_sent\": " << num_threads * msgs_per_thread
                  << ", \"total_received\": " << received
                  << ", \"elapsed_s\": " << elapsed_s
                  << ", \"aggregate_throughput\": " << received / elapsed_s
                  << ", \"avg_latency_us\": " << avg_lat << "}";
    }

    // --- Concurrent service calls ---
    {
        auto node = rclcpp::Node::make_shared("cpp_bench_conc_srv");
        auto srv = node->create_service<example_interfaces::srv::AddTwoInts>(
            "bench_conc_srv_cpp",
            [](const example_interfaces::srv::AddTwoInts::Request::SharedPtr req,
               example_interfaces::srv::AddTwoInts::Response::SharedPtr resp) {
                resp->sum = req->a + req->b;
            });

        auto executor = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
        executor->add_node(node);
        std::thread spin_thread([&executor]() { executor->spin(); });

        auto client = node->create_client<example_interfaces::srv::AddTwoInts>("bench_conc_srv_cpp");
        while (!client->wait_for_service(seconds(1))) {}

        std::atomic<int64_t> total_calls{0};
        std::mutex lat_mutex;
        std::vector<int64_t> latencies;
        latencies.reserve(num_threads * calls_per_thread);

        std::vector<std::thread> workers;
        auto t0 = high_resolution_clock::now();
        for (int t = 0; t < num_threads; ++t) {
            workers.emplace_back([&, t]() {
                for (int i = 0; i < calls_per_thread; ++i) {
                    auto req = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
                    req->a = t;
                    req->b = i;
                    auto call_t0 = high_resolution_clock::now();
                    auto future = client->async_send_request(req);
                    if (future.wait_for(seconds(5)) == std::future_status::ready) {
                        auto call_t1 = high_resolution_clock::now();
                        auto lat = duration_cast<nanoseconds>(call_t1 - call_t0).count();
                        std::lock_guard<std::mutex> lk(lat_mutex);
                        latencies.push_back(lat);
                        total_calls.fetch_add(1);
                    }
                }
            });
        }
        for (auto& w : workers) w.join();
        auto t1 = high_resolution_clock::now();
        double elapsed_s = duration<double>(t1 - t0).count();

        executor->cancel();
        spin_thread.join();

        double avg_lat = 0;
        if (!latencies.empty()) {
            double sum = std::accumulate(latencies.begin(), latencies.end(), 0.0);
            avg_lat = sum / latencies.size() / 1000.0;
        }

        std::cout << ", \"service_concurrent\": {"
                  << "\"total_calls\": " << total_calls.load()
                  << ", \"elapsed_s\": " << elapsed_s
                  << ", \"aggregate_throughput\": " << total_calls.load() / elapsed_s
                  << ", \"avg_latency_us\": " << avg_lat << "}";
    }

    // --- Concurrent param read/write ---
    {
        auto node = rclcpp::Node::make_shared("cpp_bench_conc_param");
        node->declare_parameter("conc_int", 0);

        std::atomic<int64_t> total_ops{0};
        std::vector<std::thread> workers;
        auto t0 = high_resolution_clock::now();
        for (int t = 0; t < num_threads; ++t) {
            workers.emplace_back([&node, param_ops_per_thread, &total_ops, t]() {
                for (int i = 0; i < param_ops_per_thread; ++i) {
                    node->set_parameter(rclcpp::Parameter("conc_int", t * 1000 + i));
                    volatile auto v = node->get_parameter("conc_int").as_int();
                    (void)v;
                    total_ops.fetch_add(2);
                }
            });
        }
        for (auto& w : workers) w.join();
        auto t1 = high_resolution_clock::now();
        double elapsed_s = duration<double>(t1 - t0).count();

        std::cout << ", \"param_concurrent_rw\": {"
                  << "\"total_ops\": " << total_ops.load()
                  << ", \"elapsed_s\": " << elapsed_s
                  << ", \"aggregate_throughput\": " << total_ops.load() / elapsed_s << "}";
    }

    std::cout << "}";
}

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);

    int topic_iters = 5000;
    int service_iters = 2000;
    int action_order = 20;
    int param_iters = 100000;

    int conc_threads = 4;
    int conc_msgs = 2000;
    int conc_calls = 500;
    int conc_param_ops = 10000;

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg.find("--topic-iters=") == 0) topic_iters = std::stoi(arg.substr(14));
        else if (arg.find("--service-iters=") == 0) service_iters = std::stoi(arg.substr(16));
        else if (arg.find("--action-order=") == 0) action_order = std::stoi(arg.substr(15));
        else if (arg.find("--param-iters=") == 0) param_iters = std::stoi(arg.substr(14));
        else if (arg.find("--conc-threads=") == 0) conc_threads = std::stoi(arg.substr(15));
        else if (arg.find("--conc-msgs=") == 0) conc_msgs = std::stoi(arg.substr(12));
        else if (arg.find("--conc-calls=") == 0) conc_calls = std::stoi(arg.substr(13));
        else if (arg.find("--conc-param-ops=") == 0) conc_param_ops = std::stoi(arg.substr(17));
    }

    std::cout << "{\"language\": \"cpp\"," << std::endl;

    {
        auto node = rclcpp::Node::make_shared("cpp_bench_topic");
        benchmark_topic(node, topic_iters);
        std::cout << "," << std::endl;
    }
    {
        auto node = rclcpp::Node::make_shared("cpp_bench_service");
        benchmark_service(node, service_iters);
        std::cout << "," << std::endl;
    }
    {
        auto node = rclcpp::Node::make_shared("cpp_bench_action");
        benchmark_action(node, action_order);
        std::cout << "," << std::endl;
    }
    {
        auto node = rclcpp::Node::make_shared("cpp_bench_param");
        benchmark_param(node, param_iters);
        std::cout << "," << std::endl;
    }

    benchmark_concurrency(conc_threads, conc_msgs, conc_calls, conc_param_ops);
    std::cout << std::endl;

    std::cout << "}" << std::endl;
    rclcpp::shutdown();
    return 0;
}

File: src/py_pkg/py_pkg/benchmark_node.py
import json
import sys
import time
import threading
from concurrent.futures import ThreadPoolExecutor

import rclpy
from rclpy.action import ActionClient as RclpyActionClient
from rclpy.action import ActionServer as RclpyActionServer
from rclpy.node import Node
from rclpy.executors import SingleThreadedExecutor, MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup
from std_msgs.msg import String
from example_interfaces.srv import AddTwoInts
from example_interfaces.action import Fibonacci


def percentile(sorted_data, pct):
    idx = int(len(sorted_data) * pct / 100.0)
    idx = min(idx, len(sorted_data) - 1)
    return sorted_data[idx]


def compute_stats(latencies, elapsed_s):
    latencies.sort()
    n = len(latencies)
    return {
        "avg_us": sum(latencies) / n / 1000.0,
        "min_us": latencies[0] / 1000.0,
        "max_us": latencies[-1] / 1000.0,
        "p50_us": percentile(latencies, 50) / 1000.0,
        "p95_us": percentile(latencies, 95) / 1000.0,
        "p99_us": percentile(latencies, 99) / 1000.0,
        "throughput": n / elapsed_s,
    }


def benchmark_topic(iterations):
    node = Node("py_bench_topic")
    pub = node.create_publisher(String, "bench_topic_py", 100)
    latencies = []

    def on_msg(msg):
        now_ns = time.time_ns()
        sent_ns = int(msg.data)
        latencies.append(now_ns - sent_ns)

    node.create_subscription(String, "bench_topic_py", on_msg, 100)

    start = time.time()
    for _ in range(iterations):
        msg = String()
        msg.data = str(time.time_ns())
        pub.publish(msg)
        rclpy.spin_once(node, timeout_sec=0)
        time.sleep(0.00001)

    for _ in range(50):
        rclpy.spin_once(node, timeout_sec=0.001)

    elapsed = time.time() - start
    node.destroy_node()

    if latencies:
        return compute_stats(latencies, elapsed)
    return {"error": "no messages received"}


def benchmark_service(iterations):
    node = Node("py_bench_service")

    def handle(req, resp):
        resp.sum = req.a + req.b
        return resp

    node.create_service(AddTwoInts, "bench_srv_py", handle)
    client = node.create_client(AddTwoInts, "bench_srv_py")
    while not client.wait_for_service(timeout_sec=0.1):
        rclpy.spin_once(node, timeout_sec=0)

    latencies = []
    start = time.time()
    for i in range(iterations):
        req = AddTwoInts.Request()
        req.a = i
        req.b = 1
        t0 = time.time_ns()
        future = client.call_async(req)
        rclpy.spin_until_future_complete(node, future, timeout_sec=5.0)
        t1 = time.time_ns()
        latencies.append(t1 - t0)
    elapsed = time.time() - start
    node.destroy_node()
    return compute_stats(latencies, elapsed)


def benchmark_action(fib_order):
    node = Node("py_bench_action")
    feedback_count = 0

    def execute_cb(goal_handle):
        fb = Fibonacci.Feedback()
        prev, curr = 0, 1
        for _ in range(goal_handle.request.order):
            fb.sequence.append(curr)
            goal_handle.publish_feedback(fb)
            prev, curr = curr, prev + curr
        goal_handle.succeed()
        result = Fibonacci.Result()
        result.sequence = fb.sequence
        return result

    _server = RclpyActionServer(node, Fibonacci, "bench_action_py", execute_cb)
    action_client = RclpyActionClient(node, Fibonacci, "bench_action_py")
    action_client.wait_for_server(timeout_sec=5.0)

    goal = Fibonacci.Goal()
    goal.order = fib_order

    def fb_cb(_feedback_msg):
        nonlocal feedback_count
        feedback_count += 1

    t0 = time.time_ns()
    future = action_client.send_goal_async(goal, feedback_callback=fb_cb)
    rclpy.spin_until_future_complete(node, future, timeout_sec=10.0)
    goal_handle = future.result()

    if goal_handle and goal_handle.accepted:
        result_future = goal_handle.get_result_async()
        rclpy.spin_until_future_complete(node, result_future, timeout_sec=10.0)

    t1 = time.time_ns()
    node.destroy_node()

    return {
        "total_us": (t1 - t0) / 1000.0,
        "feedback_count": feedback_count,
        "fib_order": fib_order,
    }


def benchmark_param(iterations):
    node = Node("py_bench_param")
    node.declare_parameter("bench_int", 0)
    node.declare_parameter("bench_str", "init")

    t0 = time.time()
    for i in range(iterations):
        node.set_parameters(
            [rclpy.parameter.Parameter("bench_int", value=i)]
        )
    t1 = time.time()
    write_elapsed = t1 - t0

    t2 = time.time()
    for _ in range(iterations):
        _ = node.get_parameter("bench_int").value
    t3 = time.time()
    read_elapsed = t3 - t2

    node.destroy_node()
    return {
        "write_ops_per_sec": iterations / write_elapsed,
        "read_ops_per_sec": iterations / read_elapsed,
        "iterations": iterations,
    }


def benchmark_concurrency(num_threads, msgs_per_thread, calls_per_thread,
                          param_ops_per_thread):
    conc = {"thread_count": num_threads}

    # --- Multi-publisher topic ---
    total_msgs = num_threads * msgs_per_thread
    node = Node("py_bench_conc_topic")
    latencies = []
    lat_lock = threading.Lock()
    recv_count = [0]

    from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy
    topic_qos = QoSProfile(
        history=QoSHistoryPolicy.KEEP_LAST,
        depth=total_msgs,
        reliability=QoSReliabilityPolicy.RELIABLE,
    )

    def on_msg(msg):
        now_ns = time.time_ns()
        sent_ns = int(msg.data)
        with lat_lock:
            latencies.append(now_ns - sent_ns)
            recv_count[0] += 1

    cb_group = ReentrantCallbackGroup()
    node.create_subscription(String, "bench_conc_topic_py", on_msg, topic_qos,
                             callback_group=cb_group)

    executor = MultiThreadedExecutor(num_threads=num_threads + 1)
    executor.add_node(node)
    spin_thread = threading.Thread(target=executor.spin, daemon=True)
    spin_thread.start()

    def pub_worker():
        pub = node.create_publisher(String, "bench_conc_topic_py", topic_qos)
        for _ in range(msgs_per_thread):
            msg = String()
            msg.data = str(time.time_ns())
            pub.publish(msg)

    t0 = time.time()
    with ThreadPoolExecutor(max_workers=num_threads) as pool:
        futs = [pool.submit(pub_worker) for _ in range(num_threads)]
        for f in futs:
            f.result()
    drain_start = time.time()
    while time.time() - drain_start < 5.0:
        with lat_lock:
            if recv_count[0] >= total_msgs:
                break
        time.sleep(0.01)
    t1 = time.time()
    elapsed = t1 - t0

    executor.shutdown()

    total_sent = total_msgs
    received = len(latencies)
    avg_lat = (sum(latencies) / received / 1000.0) if received else 0
    conc["topic_multi_pub"] = {
        "total_sent": total_sent,
        "total_received": received,
        "elapsed_s": round(elapsed, 4),
        "aggregate_throughput": round(received / elapsed, 2),
        "avg_latency_us": round(avg_lat, 2),
    }
    node.destroy_node()

    # --- Concurrent service calls ---
    node = Node("py_bench_conc_srv")
    cb_group = ReentrantCallbackGroup()

    def handle(req, resp):
        resp.sum = req.a + req.b
        return resp

    node.create_service(AddTwoInts, "bench_conc_srv_py", handle,
                        callback_group=cb_group)
    client = node.create_client(AddTwoInts, "bench_conc_srv_py",
                                callback_group=cb_group)

    executor = MultiThreadedExecutor(num_threads=num_threads + 1)
    executor.add_node(node)
    spin_thread = threading.Thread(target=executor.spin, daemon=True)
    spin_thread.start()

    while not client.wait_for_service(timeout_sec=0.1):
        pass

    call_latencies = []
    call_lock = threading.Lock()
    call_count = [0]

    def srv_worker(tid):
        for i in range(calls_per_thread):
            req = AddTwoInts.Request()
            req.a = tid
            req.b = i
            ct0 = time.time_ns()
            future = client.call_async(req)
            rclpy.spin_until_future_complete(node, future, timeout_sec=5.0)
            ct1 = time.time_ns()
            with call_lock:
                call_latencies.append(ct1 - ct0)
                call_count[0] += 1

    t0 = time.time()
    with ThreadPoolExecutor(max_workers=num_threads) as pool:
        futs = [pool.submit(srv_worker, t) for t in range(num_threads)]
        for f in futs:
            f.result()
    t1 = time.time()
    elapsed = t1 - t0

    executor.shutdown()

    total_calls = call_count[0]
    avg_lat = (sum(call_latencies) / total_calls / 1000.0) if total_calls else 0
    conc["service_concurrent"] = {
        "total_calls": total_calls,
        "elapsed_s": round(elapsed, 4),
        "aggregate_throughput": round(total_calls / elapsed, 2),
        "avg_latency_us": round(avg_lat, 2),
    }
    node.destroy_node()

    # --- Concurrent param read/write ---
    node = Node("py_bench_conc_param")
    node.declare_parameter("conc_int", 0)
    total_ops = [0]
    ops_lock = threading.Lock()

    def param_worker(tid):
        local_ops = 0
        for i in range(param_ops_per_thread):
            node.set_parameters(
                [rclpy.parameter.Parameter("conc_int", value=tid * 1000 + i)]
            )
            _ = node.get_parameter("conc_int").value
            local_ops += 2
        with ops_lock:
            total_ops[0] += local_ops

    t0 = time.time()
    with ThreadPoolExecutor(max_workers=num_threads) as pool:
        futs = [pool.submit(param_worker, t) for t in range(num_threads)]
        for f in futs:
            f.result()
    t1 = time.time()
    elapsed = t1 - t0

    conc["param_concurrent_rw"] = {
        "total_ops": total_ops[0],
        "elapsed_s": round(elapsed, 4),
        "aggregate_throughput": round(total_ops[0] / elapsed, 2),
    }
    node.destroy_node()

    return conc


def benchmark_multi_node(num_threads, msgs_per_thread, calls_per_thread):
    from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy
    mn = {"thread_count": num_threads}
    total_msgs = num_threads * msgs_per_thread

    topic_qos = QoSProfile(
        history=QoSHistoryPolicy.KEEP_LAST,
        depth=total_msgs,
        reliability=QoSReliabilityPolicy.RELIABLE,
    )

    # --- Multi-node topic: each pub thread owns its own node ---
    latencies = []
    lat_lock = threading.Lock()
    recv_count = [0]
    sub_ready = threading.Event()
    stop_sub = threading.Event()

    def sub_worker():
        sub_node = Node("py_mn_sub")
        sub_executor = SingleThreadedExecutor()
        sub_executor.add_node(sub_node)

        def on_msg(msg):
            now_ns = time.time_ns()
            sent_ns = int(msg.data)
            with lat_lock:
                latencies.append(now_ns - sent_ns)
                recv_count[0] += 1

        sub_node.create_subscription(String, "bench_mn_topic_py", on_msg, topic_qos)
        sub_ready.set()
        while not stop_sub.is_set():
            sub_executor.spin_once(timeout_sec=0.001)
        sub_node.destroy_node()

    sub_thread = threading.Thread(target=sub_worker, daemon=True)
    sub_thread.start()
    sub_ready.wait()
    time.sleep(0.1)

    def pub_worker(tid):
        pub_node = Node(f"py_mn_pub_{tid}")
        pub = pub_node.create_publisher(String, "bench_mn_topic_py", topic_qos)
        for _ in range(msgs_per_thread):
            msg = String()
            msg.data = str(time.time_ns())
            pub.publish(msg)
        pub_node.destroy_node()

    t0 = time.time()
    with ThreadPoolExecutor(max_workers=num_threads) as pool:
        futs = [pool.submit(pub_worker, t) for t in range(num_threads)]
        for f in futs:
            f.result()
    drain_start = time.time()
    while time.time() - drain_start < 5.0:
        with lat_lock:
            if recv_count[0] >= total_msgs:
                break
        time.sleep(0.01)
    t1 = time.time()
    elapsed = t1 - t0

    stop_sub.set()
    sub_thread.join(timeout=2)

    total_sent = total_msgs
    received = len(latencies)
    avg_lat = (sum(latencies) / received / 1000.0) if received else 0
    mn["topic"] = {
        "total_sent": total_sent,
        "total_received": received,
        "elapsed_s": round(elapsed, 4),
        "aggregate_throughput": round(received / elapsed, 2),
        "avg_latency_us": round(avg_lat, 2),
    }

    # --- Multi-node service: server in own thread, each client in own node ---
    srv_ready = threading.Event()
    stop_srv = threading.Event()

    def srv_worker():
        srv_node = Node("py_mn_srv")
        srv_executor = SingleThreadedExecutor()
        srv_executor.add_node(srv_node)

        def handle(req, resp):
            resp.sum = req.a + req.b
            return resp

        srv_node.create_service(AddTwoInts, "bench_mn_srv_py", handle)
        srv_ready.set()
        while not stop_srv.is_set():
            srv_executor.spin_once(timeout_sec=0.001)
        srv_node.destroy_node()

    srv_thread = threading.Thread(target=srv_worker, daemon=True)
    srv_thread.start()
    srv_ready.wait()
    time.sleep(0.1)

    call_latencies = []
    call_lock = threading.Lock()
    call_count = [0]

    def cli_worker(tid):
        cli_node = Node(f"py_mn_cli_{tid}")
        cli_executor = SingleThreadedExecutor()
        cli_executor.add_node(cli_node)
        client = cli_node.create_client(AddTwoInts, "bench_mn_srv_py")
        while not client.wait_for_service(timeout_sec=0.1):
            cli_executor.spin_once(timeout_sec=0)
        for i in range(calls_per_thread):
            req = AddTwoInts.Request()
            req.a = tid
            req.b = i
            ct0 = time.time_ns()
            future = client.call_async(req)
            cli_executor.spin_until_future_complete(future, timeout_sec=5.0)
            ct1 = time.time_ns()
            with call_lock:
                call_latencies.append(ct1 - ct0)
                call_count[0] += 1
        cli_node.destroy_node()

    t0 = time.time()
    with ThreadPoolExecutor(max_workers=num_threads) as pool:
        futs = [pool.submit(cli_worker, t) for t in range(num_threads)]
        for f in futs:
            f.result()
    t1 = time.time()
    elapsed = t1 - t0

    stop_srv.set()
    srv_thread.join(timeout=2)

    total_calls = call_count[0]
    avg_lat = (sum(call_latencies) / total_calls / 1000.0) if total_calls else 0
    mn["service"] = {
        "total_calls": total_calls,
        "elapsed_s": round(elapsed, 4),
        "aggregate_throughput": round(total_calls / elapsed, 2),
        "avg_latency_us": round(avg_lat, 2),
    }

    return mn


def main(args=None):
    rclpy.init(args=args)

    topic_iters = 5000
    service_iters = 2000
    action_order = 20
    param_iters = 100000
    conc_threads = 4
    conc_msgs = 2000
    conc_calls = 500
    conc_param_ops = 10000

    argv = sys.argv[1:]
    for arg in argv:
        if arg.startswith("--topic-iters="):
            topic_iters = int(arg.split("=")[1])
        elif arg.startswith("--service-iters="):
            service_iters = int(arg.split("=")[1])
        elif arg.startswith("--action-order="):
            action_order = int(arg.split("=")[1])
        elif arg.startswith("--param-iters="):
            param_iters = int(arg.split("=")[1])
        elif arg.startswith("--conc-threads="):
            conc_threads = int(arg.split("=")[1])
        elif arg.startswith("--conc-msgs="):
            conc_msgs = int(arg.split("=")[1])
        elif arg.startswith("--conc-calls="):
            conc_calls = int(arg.split("=")[1])
        elif arg.startswith("--conc-param-ops="):
            conc_param_ops = int(arg.split("=")[1])

    results = {"language": "python"}
    results["topic"] = benchmark_topic(topic_iters)
    results["service"] = benchmark_service(service_iters)
    results["action"] = benchmark_action(action_order)
    results["param"] = benchmark_param(param_iters)
    results["concurrency"] = benchmark_concurrency(
        conc_threads, conc_msgs, conc_calls, conc_param_ops)
    results["multi_node"] = benchmark_multi_node(
        conc_threads, conc_msgs, conc_calls)

    print(json.dumps(results, indent=2))
    rclpy.shutdown()
